File: 0085-maximal-rectangle/0085-maximal-rectangle.cpp
class Solution {
    
    vector<int>next_small(vector<int>&arr){
        int n=arr.size();
        stack<int>st;
        
        st.push(-1);
        
        vector<int>ans(n);
        
        for(int i=n-1;i>=0;i--){
            int curr=arr[i];
            
            while(st.top()!=-1 && arr[st.top()]>=curr){
                st.pop();
            }
            
            
            // smaller element found
            
            ans[i]=st.top();
            
            st.push(i);
        }
        
        return ans;
    }
    
    vector<int>prev_small(vector<int>&arr ){
        int n=arr.size();
        stack<int>st;
        
        st.push(-1);
        
        vector<int>ans(n);
        
        for(int i=0;i<n;i++){
            int curr=arr[i];
            
            while(st.top()!=-1 && arr[st.top()]>=curr){
                st.pop();
            }
            
            
            // smaller element found
            
            ans[i]=st.top();
            
            st.push(i);
        }
        
        return ans;
    }
    
    int largest_area(vector<int>&heights){
        int n=heights.size();
        
        vector<int>next(n);
        next=next_small(heights);
        
        vector<int>prev(n);
        prev=prev_small(heights);
        int maxi=0;
        
        for(int i=0;i<n;i++){
            int length=heights[i];
            
            if(next[i]==-1){
                next[i]=n;
            }
            int breadth=next[i]-prev[i]-1;
            
            
            int area=length*breadth;
            maxi=max(maxi,area);
            
        }
        
        return maxi;
    }
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        
        int maxi=0;
        
        vector<int>height(matrix[0].size(),0);
        
        for(int i=0;i<matrix.size();i++){
            for(int j=0;j<height.size();j++){
                if(matrix[i][j]=='1'){
                    height[j]++;
                }
                else{
                    height[j]=0;
                }
               
            }
             maxi=max(maxi,largest_area(height));
        }
        
        return maxi;
    }
};

File: 0085-maximal-rectangle/0085-maximal-rectangle.js
/**
 * @param {character[][]} matrix
 * @return {number}
 */
var maximalRectangle = function(matrix) {
    if (matrix.length === 0) return 0;
    let n = matrix[0].length;
    let height = Array(n).fill(0);
    let ans = 0;

    for (let row of matrix) {
        for (let i = 0; i < n; i++)
            height[i] = row[i] === '1' ? height[i] + 1 : 0;

        let stack = [];
        for (let i = 0; i <= n; i++) {
            let cur = i === n ? 0 : height[i];
            while (stack.length && height[stack[stack.length-1]] > cur) {
                let h = height[stack.pop()];
                let w = stack.length === 0 ? i : i - stack[stack.length-1] - 1;
                ans = Math.max(ans, h * w);
            }
            stack.push(i);
        }
    }
    return ans;
};


